core/src/main/java/org/elasticsearch/common/geo/parsers/GeoWKTParser.java

```diff
@@ -0,0 +1,279 @@
+import org.elasticsearch.common.logging.Loggers;
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.index.mapper.GeoShapeFieldMapper;
+
+import java.io.IOException;
+import java.io.StreamTokenizer;
+import java.util.List;
+
+/**
+ * Parses shape geometry represented in WKT format
+ *
+ * complies with OGC® document: 12-063r5 and ISO/IEC 13249-3:2016 standard
+ * located at http://docs.opengeospatial.org/is/12-063r5/12-063r5.html
+ */
+public class GeoWKTParser {
+    public static final String EMPTY = "EMPTY";
+    public static final String SPACE = Loggers.SPACE;
+    public static final String LPAREN = "(";
+    public static final String RPAREN = ")";
+    public static final String COMMA = ",";
+    private static final String NAN = "NaN";
+
+    private static final String NUMBER = "<NUMBER>";
+    private static final String EOF = "END-OF-STREAM";
+    private static final String EOL = "END-OF-LINE";
+
+    // no instance
+    private GeoWKTParser() {}
+
+    public static ShapeBuilder parse(XContentParser parser, GeoShapeFieldMapper shapeFieldMapper)
+            throws IOException, ElasticsearchParseException {
+        FastStringReader reader = new FastStringReader(parser.text());
+        try {
+            // setup the tokenizer; configured to read words w/o numbers
+            StreamTokenizer tokenizer = new StreamTokenizer(reader);
+            tokenizer.resetSyntax();
+            tokenizer.wordChars('a', 'z');
+            tokenizer.wordChars('A', 'Z');
+            tokenizer.wordChars(128 + 32, 255);
+            tokenizer.wordChars('0', '9');
+            tokenizer.wordChars('-', '-');
+            tokenizer.wordChars('+', '+');
+            tokenizer.wordChars('.', '.');
+            tokenizer.whitespaceChars(0, ' ');
+            tokenizer.commentChar('#');
+            return parseGeometry(tokenizer);
+        } finally {
+            reader.close();
+        }
+    }
+
+    /** parse geometry from the stream tokenizer */
+    private static ShapeBuilder parseGeometry(StreamTokenizer stream) throws IOException, ElasticsearchParseException {
+        final GeoShapeType type = GeoShapeType.forName(nextWord(stream));
+        switch (type) {
+            case POINT:
+                return parsePoint(stream);
+            case MULTIPOINT:
+                return parseMultiPoint(stream);
+            case LINESTRING:
+                return parseLine(stream);
+            case MULTILINESTRING:
+                return parseMultiLine(stream);
+            case POLYGON:
+                return parsePolygon(stream);
+            case MULTIPOLYGON:
+                return parseMultiPolygon(stream);
+            case ENVELOPE:
+                return parseBBox(stream);
+        }
+        throw new IllegalArgumentException("Unknown geometry type: " + type);
+    }
+
+    private static EnvelopeBuilder parseBBox(StreamTokenizer stream) throws IOException, ElasticsearchParseException {
```

Comment:
This will also accept something like "BBOX (-10) 10) -10) 10)", I think this should throw an error. Its a bit constructed but I think the parser shouldn't be that lenient.

Reply:
:+1: